### DISCUSSION TITLE
#10 📑 Weekly Report 10

### DISCUSSION THREAD

**Discussion by fehlend0** [SE-TINF22B6/Statify#83] [Blog]
Hello everyone👋
The week is almost over, so we want to give you an update on our project.

## Updated data model:

We updated our data model so that it allows for tracking and analyzing user interactions with music, including their favorite artists, tracks, playlists, and associated statistics over time.

![DataModel](https://github.com/SE-TINF22B6/Statify/assets/106541016/488490e6-bfc1-49e8-8391-7b049c75bd95)

1. User Table:
   - Stores information about users.
   - Includes a unique identifier (GUID) for each user (`Guid uniqueidentifier`).
   - Requires a non-null and unique access token (`access_token varchar(max) NOT NULL UNIQUE`).

2. TopArtistsStats Table:
   - Stores statistics about top artists for users within a specified time period.
   - Includes a primary key (`PK Guid uniqueidentifier`) and foreign keys (`FK user_guid uniqueidentifier`, FK first_artist, FK second_artist, FK third_artist, FK fourth_artist, `FK fifth_artist`) to link to the User and Artist tables.
   - Tracks start and end dates for the statistics (`start_date date NOT NULL`, `end_date date NOT NULL`).

3. Artist Table:
   - Stores information about artists.
   - Includes a unique identifier (GUID) for each artist (`PK Guid uniqueidentifier`).
   - Stores the name of the artist (`name varchar(max) NOT NULL`), and optionally an image URL (`image_url varchar(max)`).

4. PlaylistStats Table:
   - Stores statistics about user playlists.
   - Includes a primary key (`PK Guid uniqueidentifier`) and a foreign key (`FK user_guid uniqueidentifier`) to link to the User table.
   - Tracks playlist name (`name varchar(max) NOT NULL`), number of tracks (`tracks_number int NOT NULL`), duration (`duration_ms int NOT NULL`), top genre (`top_genre varchar(max) NOT NULL`), and related statistics.

5. TopTracksStats Table:
   - Stores statistics about top tracks for users within a specified time period.
   - Similar to TopArtistsStats but for tracks.
   - Includes a primary key (`PK Guid uniqueidentifier`) and foreign keys (`FK user_guid uniqueidentifier`, FK first_track, FK second_track, FK third_track, FK fourth_track, `FK fifth_track`) to link to the User and Track tables.

6. Track Table:
   - Stores information about tracks.
   - Includes a unique identifier (GUID) for each track (`PK Guid uniqueidentifier`).
   - Stores the name of the track (`name varchar(max) NOT NULL`), the name of the artist (`artist_name varchar(max) NOT NULL`), and optionally an image URL (`image_url varchar(max)`).

7. AudioFeatures Table:
   - Stores audio features for tracks (e.g., tempo, danceability, energy).
   - Includes a primary key (`PK Guid uniqueidentifier`).
   - Stores various audio features (`duration_ms int`, key int, tempo float, mode int, danceability float, energy float, loudness float, speechiness float, acousticness float, instrumentalness float, liveness float, `valence float`).

## Set up a MySQL Server:

A virtual machine was created on bwCloud, where MySQL Server was installed and necessary network configurations were made. A database was created based on an updated data model, and a connection to the MySQL Server was established within the application.

## Updates in the frontend:

![MicrosoftTeams-image (2)](https://github.com/SE-TINF22B6/Statify/assets/106541016/d2b090be-4e6e-4f35-b85f-03169f6df599)
![MicrosoftTeams-image (1)](https://github.com/SE-TINF22B6/Statify/assets/106541016/a6952ded-7353-4165-bd96-51166ead5a1e)

We have added the page for statistics (currently with sample data). The "box" with the statistics consists of several components: StatisticsFrame includes the border, heading, date, and separator. The individual rows/artists are StatisticItem components. In the last screenshot, there is also the NumberItem component, which includes a title (e.g., Top Genre), subtitle (e.g., Pop), number, and unit (e.g., Songs). The login with Spotify has been implemented correctly. When you click on "Get Started" or "Log in," a request is sent to the backend, and you are redirected to the Spotify login page and then to the statistics page. Some minor styling adjustments/improvements have also been made.

Stay tuned for more to come!

**Comment by giomskii** [comment]
Hello Team Statify,
it's nice to hear that you have established your database on a server and that the signin via Spotify works, since it is very crucial for your application. The data models are also looking quite clean and seem to be decomposed just in the right way. Are you going to use a REST framework which handles the data transfer between frontend and database or are you doing it in a different way?
Nonetheless, we can't wait to see your progress in the next weeks! - Team DHBWhub
